kong/client_test.go

```diff
@@ -205,3 +207,123 @@ func TestBaseRootURL(t *testing.T) {
 		require.Equal(t, client.BaseRootURL(), "https://customkong2.com")
 	})
 }
+
+func TestReloadDeclarativeRawConfig(t *testing.T) {
+	RunWhenDBMode(t, "off")
+
+	tests := []struct {
+		name    string
+		config  Configuration
+		wantErr bool
+	}{
+		{
+			name: "basic config works",
+			config: Configuration{
+				"_format_version": "1.1",
+				"services": []Configuration{
+					{
+						"host":     "mockbin.com",
+						"port":     443,
+						"protocol": "https",
+						"routes": []Configuration{
+							{"paths": []string{"/"}},
+						},
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "missing _format_version fails",
+			config: Configuration{
+				"services": []Configuration{
+					{
+						"host":     "mockbin.com",
+						"port":     443,
+						"protocol": "https",
+						"routes": []Configuration{
+							{"paths": []string{"/"}},
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "invalid config fails",
+			config: Configuration{
+				"dummy_key": []Configuration{
+					{
+						"host":     "mockbin.com",
+						"port":     443,
+						"protocol": "https",
+					},
+				},
+			},
+			wantErr: true,
+		},
+	}
+
+	// POST /config actually interprets query string fields as additional fields within the config unless they're
+	// explicitly stripped by the API code (no idea why this behavior exists). Without stripping "flatten_errors" in
+	// https://github.com/Kong/kong/blob/master/kong/api/routes/config.lua#L71 Kong will actually reject configuration
+	// because it thinks "flatten_errors" is an actual (and invalid) field inside the config.
+	//
+	// This is the one test where we want version-dependent behavior, but only for one value within the test. We test
+	// config updates on all DB-less capable versions, but only set flattenErrors=true on 3.2+. To handle that, this
+	// snippet is borrowed from RunWhenKong, to allow toggling that behavior only on or off depending on the version.
+	var flattenErrors bool
+	client, err := NewTestClient(nil, nil)
+	if err != nil {
+		t.Error(err)
+	}
+	info, err := client.Root(defaultCtx)
+	if err != nil {
+		t.Error(err)
+	}
+	version := VersionFromInfo(info)
+	currentVersion, err := ParseSemanticVersion(version)
+	if err != nil {
+		t.Error(err)
+	}
+	r, err := NewRange(">=3.2.0")
+	if err != nil {
+		t.Error(err)
+	}
+	if r(currentVersion) {
+		t.Log("Kong version >=3.2, enabling flattenErrors for ReloadDeclarativeRawConfig")
+		flattenErrors = true
+	} else {
+		t.Log("Kong version <3.2, disabling flattenErrors for ReloadDeclarativeRawConfig")
+	}
```

Comment:
Just a thought that instead of doing all these shenanigans we could just add a `func()` field in table driven test struct which will call appropriate `RunWhenKong(T, ">=3.2")` etc making that version filtering be bound to a concrete test case e.g.

When done this way a potential test failure will be much more specific (subjectively ?)

Reply:
Not quite sure I follow, can you show a code example or add a commit?

The other thing I'd considered was making the go-kong test helper suite support a second set of lighter `IfKongVersion(">=3.2")` helpers that did not interact with the test object directly (hey! same advice from the other comment that we aren't following yet :laughing: ), but I figured that, if I was going to build that out, it should have support across the board for the current whole-test helpers, so `IfKongEnterprise(...)` also and so on. So long as this is the odd test out, I YAGNIed on that and instead brought the version checking logic under this one oddball test.